csharp/src/SeedLang/Block/InlineTextParser.cs

```diff
@@ -28,62 +29,72 @@ internal class InlineTextParser : BaseParser {
     // The listener interface to be notified when the tokens of an expression inline text are
     // visited.
     internal interface IInlineTextListener {
-      void VisitArithmeticOperator(string op);
-      void VisitIdentifier(string name);
-      void VisitNumber(string number);
-      void VisitString(string str);
-      void VisitOpenParen();
-      void VisitCloseParen();
+      void VisitArithmeticOperator(string op, TextRange range);
+      void VisitIdentifier(string name, TextRange range);
+      void VisitNumber(string number, TextRange range);
+      void VisitString(string str, TextRange range);
+      void VisitOpenParen(TextRange range);
+      void VisitCloseParen(TextRange range);
+      void VisitInvalidToken(TextRange range);
     }
 
     // Visits an inline text of block programs. The given listener is notified when each token of
     // the inline text is visited. The negative sign token will be combined with the following
     // number to form a negative number.
+    //
+    // The input can be either a valid inline text or an invalid inline text. If the text is
+    // invalid, this method will still try to parse it and notify the listener about both valid and
+    // invalid tokens.
     internal void VisitInlineText(string text, IInlineTextListener listener) {
-      if (!Validate(text, "", ParseRule.Expression, null)) {
-        return;
-      }
       Lexer lexer = SetupLexer(text);
       int lastTokenType = SeedBlockParser.UNKNOWN_CHAR;
       bool negative = false;
+      IToken negativeToken = null;
       foreach (var token in lexer.GetAllTokens()) {
         if (negative && token.Type != SeedBlockLexer.NUMBER) {
           // TODO: define a const string for the negative sign?
-          listener.VisitArithmeticOperator("-");
+          listener.VisitArithmeticOperator("-", TextRange.RangeOfToken(negativeToken));
           negative = false;
         }
         switch (token.Type) {
           case SeedBlockLexer.ADD:
           case SeedBlockLexer.MUL:
           case SeedBlockLexer.DIV:
-            listener.VisitArithmeticOperator(token.Text);
+            listener.VisitArithmeticOperator(token.Text, TextRange.RangeOfToken(token));
             break;
           case SeedBlockLexer.SUB:
             if (lastTokenType == SeedBlockParser.NUMBER ||
                 lastTokenType == SeedBlockParser.CLOSE_PAREN) {
-              listener.VisitArithmeticOperator(token.Text);
+              listener.VisitArithmeticOperator(token.Text, TextRange.RangeOfToken(token));
             } else {
               negative = true;
+              negativeToken = token;
```

Comment:
negative and nagativeToken are duplicated. Is it better to remove negative variable and check if negativeToken is not null to see if met a negative sign?

Reply:
Done